### PR TITLE
RPE-284: Retrieve letters

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Create a task to alert on unprinted docs](https://tools.hmcts.net/jira/browse/RPE-284)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterMapperFactory.java [hunk 1]
```diff
@@ -14,18 +15,19 @@
...
 
-            if (timestamp == null) {
-                return null;
-            }
+        return timestamp.toInstant().atZone(ZoneOffset.UTC);
+    }
 
-            return timestamp.toInstant().atZone(ZoneOffset.UTC);
-        }
+    private static final class LetterStatusMapper implements RowMapper<LetterStatus> {
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterMapperFactory.java]: this whole block from line 18 looks weird. Basically it's just moving method outside of the inner class

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java [hunk 2]
```diff
@@ -120,6 +122,20 @@ public int updateIsFailed(UUID id) {
         );
     }
 
+    /**
+     * Retrieve a list of letters which are not failed but still not printed.
+     *
+     * @return list of not printed letters until yesterday's 5pm
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: judging by the query it's a list of letters not printed until now, that were created before yesterday 5pm.

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java [hunk 3]
```diff
@@ -120,6 +122,20 @@ public int updateIsFailed(UUID id) {
...
+     */
+    public List<NotPrintedLetter> getNotPrintedUntilYesterday() {
+        return jdbcTemplate.query(
+            "SELECT id, message_id, service, type, created_at, sent_to_print_at "
+                + "FROM letters "
+                + "WHERE is_failed = false AND printed_at IS NULL AND sent_to_print_at IS NOT NULL"
+                + "  AND sent_to_print_at < (CURRENT_DATE - integer '1' + time '17:00')",
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: it would be nice to have each `AND` in a new line imho

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java [hunk 4]
```diff
@@ -120,6 +122,20 @@ public int updateIsFailed(UUID id) {
...
+    /**
+     * Retrieve a list of letters which are not failed but still not printed.
+     *
+     * @return list of not printed letters until yesterday's 5pm
+     */
+    public List<NotPrintedLetter> getNotPrintedUntilYesterday() {
+        return jdbcTemplate.query(
+            "SELECT id, message_id, service, type, created_at, sent_to_print_at "
+                + "FROM letters "
+                + "WHERE is_failed = false AND printed_at IS NULL AND sent_to_print_at IS NOT NULL"
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: One query here we have defined column type of `is_failed` as BOOLEAN type but haven't specified default value as FALSE.I imagine it will null in that case and above query needs to be updated?

File: src/test/java/uk/gov/hmcts/reform/sendletter/data/LetterRepositoryTest.java [hunk 5]
```diff
@@ -146,4 +149,26 @@ public void should_return_empty_optional_case_when_no_letters_found() {
 
         assertThat(result.orElse(null)).isNull();
     }
+
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/data/LetterRepositoryTest.java]: Could you also add a test case where there are zero not printed letters?

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java (hunk 2 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: * that were sent to print yesterday before 5pm

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java (hunk 3 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: there are a lot nice things. new line ANDs are one of them

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java (hunk 4 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: As per migration script column does not have a default value defined. So it is a boolean false. There will be a functional test proving that once plugged all this into service. For now can "trust" it is stored as `false` from this [test](https://github.com/hmcts/send-letter-producer-service/blob/master/src/functionalTest/java/uk/gov/hmcts/reform/sendletter/controllers/GetLetterStatusTest.java#L76-L88)

File: src/test/java/uk/gov/hmcts/reform/sendletter/data/LetterRepositoryTest.java (hunk 5 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/sendletter/data/LetterRepositoryTest.java]: Was thinking about it but in a way - does really make sense? Will also cover that case with functional test

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java [hunk 6]
```diff
@@ -125,13 +125,15 @@ public int updateIsFailed(UUID id) {
     /**
      * Retrieve a list of letters which are not failed but still not printed.
      *
-     * @return list of not printed letters until yesterday's 5pm
+     * @return list of not printed letters but where sent to before yesterday's 5pm
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: typo in `were`

File: src/test/java/uk/gov/hmcts/reform/sendletter/data/LetterRepositoryTest.java (hunk 5 above)
Comment (nitinprabhu) [reply src/test/java/uk/gov/hmcts/reform/sendletter/data/LetterRepositoryTest.java]: If you are going to add functional test case then I think that is fine 👍

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java [hunk 7]
```diff
@@ -120,6 +122,22 @@ public int updateIsFailed(UUID id) {
         );
     }
 
+    /**
+     * Retrieve a list of letters which are not failed but still not printed.
+     *
+     * @return a list of unprinted letters which were sent before yesterdays 5pm deadline
+     */
+    public List<NotPrintedLetter> getNotPrintedUntilYesterday() {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: method name still wrong

File: src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java [hunk 8]
```diff
@@ -120,6 +122,22 @@ public int updateIsFailed(UUID id) {
...
+     * @return a list of unprinted letters which were sent before yesterdays 5pm deadline
+     */
+    public List<NotPrintedLetter> getStaleLetters() {
+        return jdbcTemplate.query(
+            "SELECT id, message_id, service, type, created_at, sent_to_print_at "
+                + "FROM letters "
+                + "WHERE is_failed = false"
+                + "  AND printed_at IS NULL"
+                + "  AND sent_to_print_at IS NOT NULL"
+                + "  AND sent_to_print_at < (CURRENT_DATE - integer '1' + time '17:00')",
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: Do you think it will be good to put `15:00` in `application.yml` rather than hardcoding in code say tomorrow timing changes?

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: An idea, yes. Could strip `5pm` from return description too. But then why not configure the whole thing and make it more complicated. on the other hand, it is related (I guess) with ftp availability which is defined in consumer.

I would prefer to leave this for a feature extension

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/data/LetterRepository.java]: perhaps we could just specify a number of hours that passed since the letter was sent to print
